executor/src/witgen/query_processor.rs

```diff
@@ -65,105 +70,95 @@ where
-            }
-            Expression::FunctionCall(FunctionCall { id, arguments }) => {
-                let [arg] = &arguments[..] else {
-                    return Err(IncompleteCause::ExpressionEvaluationUnimplemented(
-                        expr.to_string(),
-                    ));
-                };
-                let arg = self.evaluate_expression(arg, rows)?;
-                self.evaluate_poly(id.to_string().as_str(), arg.to_degree(), rows)
-            }
-            e => Err(IncompleteCause::ExpressionEvaluationUnimplemented(
-                e.to_string(),
-            )),
-        }
-    }
+#[derive(Clone)]
+struct Symbols<'a, T: FieldElement> {
+    // TODO we should also provide access to non-column symbols.
+    fixed_data: &'a FixedData<'a, T>,
+    rows: &'a RowPair<'a, 'a, T>,
+}
 
-    fn evaluate_poly(
+impl<'a, T: FieldElement> SymbolLookup<'a, T, Reference> for Symbols<'a, T> {
+    fn lookup(&self, name: &str) -> Result<Value<'a, T, Reference>, EvalError> {
+        let poly_id = self.fixed_data.column_by_name(name);
+        Ok(Value::Custom(Reference { poly_id }))
+    }
+    fn eval_function_application(
         &self,
-        poly: &str,
-        row: DegreeType,
-        rows: &RowPair<T>,
-    ) -> Result<T, IncompleteCause<&'a AlgebraicReference>> {
-        let poly_id = self.fixed_data.column_by_name(poly);
-        match poly_id.ptype {
+        function: Reference,
+        arguments: &[Rc<Value<'a, T, Reference>>],
+    ) -> Result<Value<'a, T, Reference>, EvalError> {
+        if arguments.len() != 1 {
+            Err(EvalError::TypeError(format!(
+                "Expected one argument, but got {}",
+                arguments.len()
+            )))?
+        };
+        let Value::Number(row) = arguments[0].as_ref() else {
+            return Err(EvalError::TypeError(format!(
+                "Expected number but got {}",
+                arguments[0]
+            )));
+        };
+        Ok(Value::Number(match function.poly_id.ptype {
             PolynomialType::Committed | PolynomialType::Intermediate => {
-                let next = if row == rows.current_row_index {
-                    false
-                } else if row == rows.current_row_index + 1 {
-                    true
-                } else {
-                    return Err(IncompleteCause::ExpressionEvaluationUnimplemented(
-                        "Referenced witness column outside of evaluation window.".to_string(),
-                    ));
-                };
-
+                let next =
+                    self.rows
+                        .row_number_next(row.to_degree())
+                        .ok_or(EvalError::OutOfBounds(format!(
+                            "Referenced row outside of window: {row}"
+                        )))?;
                 let poly_ref = AlgebraicReference {
-                    name: poly.to_string(),
-                    poly_id,
+                    name: String::new(),
```

Comment:
The lifetime of this object is just for the value lookup inside `self.rows`. I didn't want to store references inside the value returned by `lookup` because that's a mess. We could store a String, but it's really not necessary. Another way would be to provide an interface in `self.rows` that works by `poly_id`. But I can also just store an AlgebraicReference. Yes, it's only a matter of performance.